### PR TITLE
[TRTLLM-15498][perf] prepare Kimi K3 prefill metadata

### PR DESCRIPTION
**Pull request by jiaganc**
## Description

This PR adds Kimi K3-specific prefill metadata preparation:

- Make the Mamba metadata implementation used by `AttentionMetadata` overridable while retaining a lazy `Mamba2Metadata` default to avoid an import cycle.
- Select `KimiK3AttentionMetadata` for KimiLinear models using the TRTLLM attention backend.
- Build and reuse FLA-compatible 64-token KDA chunk indices on CUDA during metadata preparation, and record aligned-varlen and single-sequence hints for optimized prefill dispatch.

## Test Coverage

- Added `test_kimi_k3_mamba_metadata.py` to verify that KDA chunk indices match FLA and that metadata is refreshed and cleared correctly across executor steps.
- Registered the new test in the B200 L0 test list.

## PR Checklist

Please review the following before submitting this PR:
- PR description clearly explains what and why.
- PR follows the TRT-LLM coding guidelines to the best of my knowledge.
- Test cases cover the new code paths.
- No public API or dependency changes are introduced.
- Documentation, CODEOWNERS, and architecture diagram updates are not required for this internal optimization.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, comment `/bot help`.

## Dev Engineer Review

- Added lazy, overridable Mamba metadata construction through `AttentionMetadata.mamba_metadata_cls`.
- Preserved `Mamba2Metadata` as the default.
- Added model hooks for Kimi K3 metadata selection.
- Added reusable CUDA KDA chunk indices with 64-token FLA-compatible chunks.
- Added aligned-varlen and single-sequence prefill metadata.
- Added validation for nonpositive context lengths.
- Updated the B200 test-list entry to `test_kda_mamba_metadata.py`.
- CI reports show failures or instability in pipelines `#56948`, `#56983`, and `#57239`. Follow-up investigation and a rerun are required. Multi-GPU reruns require the `ci: full pre-merge approved` label.

## QA Engineer Review

- Added `test_k3_prepare_metadata_match_chunk_indices`.
- Added `test_prepare_refreshes_and_clears_kda_prefill_metadata`.
- Added `test_prepare_rejects_nonpositive_context_length`.
- Added `test_prepare_ignores_generation_lengths`.
- Registered the test in `tests/integration/test_lists/test-db/l0_b200.yml`.
- Coverage includes FLA chunk-index comparison, metadata refresh and clearing, invalid-input handling, and exclusion of generation lengths from prefill calculations.
- Verdict: needs follow-up because CI runs failed or were unstable.

### CODE REVIEW

**Review by tburt-nv — APPROVED** [review]

**Review by Shixiaowei02 — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 1]
```diff
@@ -3213,6 +3213,15 @@ def _set_up_attn_metadata(
             assert self.attn_metadata.kv_cache_manager is kv_cache_manager
             return self.attn_metadata
 
+        if is_kimi_linear(config):
+            if metadata_cls is not TrtllmAttentionMetadata:
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: Identity check rejects legitimate members of the TRTLLM metadata family. With a sparse-attention config, `self.attn_backend.Metadata` is `RocketTrtllmAttentionMetadata` or `MiniMaxM3MsaSparseAttentionMetadata` — both subclasses of `TrtllmAttentionMetadata` — so this raises "Kimi K3 requires the TRTLLM attention backend" when the backend *is* TRTLLM. `_run_attention_warmup` (line 1800) does the same family test with `issubclass`.

Swapping the class also discards whichever subclass the backend selected. Since `mamba_metadata_cls` is `init=False` and is only read lazily inside `_prepare_mamba_metadata`, you can set it on the constructed instance instead and drop `KimiK3AttentionMetadata` entirely:

```python
self.attn_metadata = metadata_cls(...)
if is_kimi_linear(config):
    if not issubclass(metadata_cls, TrtllmAttentionMetadata):
        raise ValueError(
            f"Kimi K3 requires the TRTLLM attention backend, got {metadata_cls.__name__}")
    self.attn_metadata.mamba_metadata_cls = KimiK3MambaMetadata
```

Either way, please include the actual class name in the message — "requires the TRTLLM attention backend" is hard to act on when the user believes they selected it.

File: tests/unittest/_torch/modules/kimi_kda/test_kimi_k3_mamba_metadata.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        num_ctx_tokens=sum(context_lengths),
+        kv_cache_manager=None,
+        request_ids=None,
+        kv_cache_params=SimpleNamespace(
+            num_cached_tokens_per_seq=torch.zeros(num_contexts, dtype=torch.int),
+        ),
+    )
+
+
+def test_k3_prepare_metadata_match_chunk_indices() -> None:
```
Comment (brnguyen2) [inline tests/unittest/_torch/modules/kimi_kda/test_kimi_k3_mamba_metadata.py:36]: `pytest.importorskip("fla")` makes the parity assertion — the part of this test that actually pins the FLA contract — vanish silently if the optional package isn't installed, leaving only the alignment-flag checks. The repo vendors an equivalent at `tensorrt_llm._torch.modules.fla.index.prepare_chunk_indices` (same `[indices.eq(0).cumsum(0) - 1, indices]` layout, same `.to(cu_seqlens)` dtype), which is always importable. Using that as the reference removes the skip and keeps the parity guarantee enforced in every environment.

File: tests/unittest/_torch/modules/kimi_kda/test_kimi_k3_mamba_metadata.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+import pytest
+import torch
+
+from tensorrt_llm._torch.modules.kimi_kda.kimi_k3_mamba_metadata import KimiK3MambaMetadata
+
+pytestmark = pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA required")
+
+
+def _make_attention_metadata(context_lengths: list[int]) -> SimpleNamespace:
```
Comment (brnguyen2) [inline tests/unittest/_torch/modules/kimi_kda/test_kimi_k3_mamba_metadata.py]: All cases are pure-context batches (`seq_lens` length == `num_contexts`). The slicing in `prepare()` (`seq_lens[:num_contexts]`, `seq_lens_cuda[:len(context_lengths)]`) is exactly what a mixed prefill+decode batch would expose if it ever regressed — decode entries sit past `num_contexts` in `seq_lens` and must not contribute chunks. Add one case where `_make_attention_metadata` takes e.g. contexts `[128, 300]` plus two decode entries of length 1, and assert `kda_chunk_indices` is unchanged from the contexts-only case (and that `kda_single_sequence_length` is still `None`).

File: tensorrt_llm/_torch/modules/kimi_kda/kimi_k3_mamba_metadata.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+        chunk_counts = [
+            (length + KDA_PREFILL_CHUNK_SIZE - 1) // KDA_PREFILL_CHUNK_SIZE
+            for length in context_lengths
+        ]
+        num_chunks = sum(chunk_counts)
+        if self._kda_chunk_indices.shape[0] < num_chunks:
+            self._kda_chunk_indices = torch.empty((num_chunks, 2), dtype=torch.long, device="cuda")
+        self.kda_chunk_indices = self._kda_chunk_indices[:num_chunks]
+        if num_chunks == 0:
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/modules/kimi_kda/kimi_k3_mamba_metadata.py]: `num_chunks == 0` is unreachable here: the early return above already handled the empty-context case, and any context length >= 1 contributes at least one chunk. If it's defensive against a zero-length context entry, note that this branch leaves `kda_chunk_indices` as a 0-row view rather than `None`, which is a different signal to `_kda_kernels` than the empty-batch path sets. Either drop the branch or make it set `None` for consistency.

**Review by brnguyen2 — APPROVED** [review]
Approving — the comments below are optional touch-ups, not blockers.

The metadata build itself looks right: dtype and layout match `fla.prepare_chunk_indices` (int64 `[seq_idx, local_chunk_idx]`, per-sequence chunk counts), the `output_size` trick keeps `repeat_interleave` from syncing, and the sliced reuse buffer still satisfies the `chunk_indices.shape[0] < 4` fallback gate in `_kda_kernels.py`. My concerns are around the dispatch and the test's skip behavior — details inline.

Two things at the PR level:

1. The new `ValueError` for KimiLinear on a non-TRTLLM backend is a behavior change the description doesn't mention. Before this PR, such a run would fall back (the mixer reads the KDA fields via `getattr` and `_kda_kernels` recomputes chunk indices from `fla`); now engine setup fails outright. If that combination was already broken for other reasons, say so in the description and keep the error; if not, a `logger.warning` plus leaving the default metadata class in place preserves the fallback.

2. For a `[perf]`-tagged change, there are no numbers. The stated win is removing a per-layer `cu_seqlens` readback during prefill — a before/after on a KDA prefill benchmark (ideally at a prefill length where the per-layer cost is visible) would substantiate it and give the next person a baseline. Also worth confirming the reported `L0_MergeRequest_PR` failures are unrelated before merge.

Minor: the commit message is tagged `[None]` while the PR title carries `TRTLLM-15498`; worth aligning so the ticket is greppable from git history.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 1 above)
Comment (jiaganc) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Addressed across 2e75adc258 through 0b1fa48c50. The engine now preserves the backend-selected metadata subclass and reads an optional model-declared `mamba_metadata_cls` field. It no longer swaps metadata classes, hardcodes Kimi, or adds a duplicate TRTLLM backend check.

File: tests/unittest/_torch/modules/kimi_kda/test_kimi_k3_mamba_metadata.py (hunk 2 above)
Comment (jiaganc) [reply tests/unittest/_torch/modules/kimi_kda/test_kimi_k3_mamba_metadata.py:36]: Addressed in 1eb8547590. The parity test now imports `tensorrt_llm._torch.modules.fla.index.prepare_chunk_indices`, so the assertion no longer skips when the external FLA package is absent.

File: tests/unittest/_torch/modules/kimi_kda/test_kimi_k3_mamba_metadata.py (hunk 3 above)
Comment (jiaganc) [reply tests/unittest/_torch/modules/kimi_kda/test_kimi_k3_mamba_metadata.py]: Addressed in 2e75adc258 and completed in 06ae2edf98. The mixed-batch test supplies two context entries followed by two decode entries, verifies the exact context-only chunk indices, and checks that `kda_single_sequence_length` remains `None`.

File: tensorrt_llm/_torch/modules/kimi_kda/kimi_k3_mamba_metadata.py (hunk 4 above)
Comment (jiaganc) [reply tensorrt_llm/_torch/modules/kimi_kda/kimi_k3_mamba_metadata.py]: Addressed in 1eb8547590. The unreachable zero-chunk branch was removed, and nonpositive context lengths are asserted as an internal scheduler invariant.

File: tensorrt_llm/_torch/modules/kimi_kda/kimi_k3_mamba_metadata.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        self._kda_chunk_indices = torch.empty((0, 2), dtype=torch.long, device="cuda")
+        self.kda_chunk_indices: torch.Tensor | None = None
+        self.kda_varlen_is_aligned: bool | None = None
+        self.kda_single_sequence_length: int | None = None
+
+    def prepare(self, attn_metadata: AttentionMetadata) -> None:
+        context_lengths = attn_metadata.seq_lens[: attn_metadata.num_contexts].tolist()
+        assert all(length > 0 for length in context_lengths), (
+            f"Expected positive context lengths, got {context_lengths}"
+        )
```
[the remark this answers is not in the document]
Comment (jiaganc) [reply tensorrt_llm/_torch/modules/kimi_kda/kimi_k3_mamba_metadata.py:55]: The assertion is intentional here because positive context lengths are an internal scheduler invariant, not user-input validation. A violation indicates a programming error, and the test explicitly covers that contract, so we are keeping the assertion.

**Review by PerkzZheng — APPROVED** [review]

**Review by tongyuantongyu — APPROVED** [review]

**Review by cascade812 — APPROVED** [review]

File: tensorrt_llm/_torch/attention_backend/interface.py [hunk 6]
```diff
@@ -43,6 +44,13 @@
         ALLOWED_LAYER_TYPES as ALLOWED_ATTENTION_LAYER_TYPES
 
 
+def _default_mamba_metadata_cls() -> Type["Mamba2Metadata"]:
```
Comment (pengbowang-nv) [inline tensorrt_llm/_torch/attention_backend/interface.py]: I think it might be better to change the design this way: inject mamba metadata class in model engine as we have done in this PR. However we have a dedicated `resolve_mamba_metadata_cls` function which will handle extracting from model and fallback. By doing so we can be more extensible and avoid these circular imports.

**Review by BowenFu — APPROVED** [review]

File: tensorrt_llm/_torch/modules/kimi_kda/kimi_k3_mamba_metadata.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            self.kda_chunk_indices = None
+            return
+
+        chunk_counts = [
+            (length + KDA_PREFILL_CHUNK_SIZE - 1) // KDA_PREFILL_CHUNK_SIZE
+            for length in context_lengths
+        ]
+        num_chunks = sum(chunk_counts)
+        if self._kda_chunk_indices.shape[0] < num_chunks:
+            self._kda_chunk_indices = torch.empty((num_chunks, 2), dtype=torch.long, device="cuda")
```
Comment (pengbowang-nv) [inline tensorrt_llm/_torch/modules/kimi_kda/kimi_k3_mamba_metadata.py]: Quick Question: Will this reallocate break cudagraph or similar graph capturing (or if such support is not in plan for now)?

File: tests/unittest/_torch/modules/kimi_kda/test_kimi_k3_mamba_metadata.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        metadata.prepare(_make_attention_metadata(context_lengths))
+
+        assert metadata.kda_varlen_is_aligned is expected_alignment
+        assert metadata.kda_single_sequence_length == expected_single_length
+        cu_seqlens = metadata.query_start_loc_long[: len(context_lengths) + 1]
+        legacy_chunk_indices = prepare_chunk_indices(cu_seqlens, chunk_size=64)
+        torch.testing.assert_close(metadata.kda_chunk_indices, legacy_chunk_indices)
+
+
+def test_prepare_refreshes_and_clears_kda_prefill_metadata() -> None:
```
Comment (pengbowang-nv) [inline tests/unittest/_torch/modules/kimi_kda/test_kimi_k3_mamba_metadata.py]: I think `test_prepare_refreshes_and_clears_kda_prefill_metadata`, `test_prepare_rejects_nonpositive_context_length` and `test_prepare_ignores_generation_lengths` may not be necessary as they are just restatement of code?

File: tensorrt_llm/_torch/attention_backend/interface.py (hunk 6 above)
Comment (jiaganc) [reply tensorrt_llm/_torch/attention_backend/interface.py]: Addressed in 640ffce03e. `resolve_mamba_metadata_cls` now lives in `model_engine.py`, reads the model class field, and falls back to `Mamba2Metadata`. `AttentionMetadata` no longer imports the concrete implementation at runtime; the model engine injects the resolved class.

File: tensorrt_llm/_torch/modules/kimi_kda/kimi_k3_mamba_metadata.py (hunk 7 above)
Comment (jiaganc) [reply tensorrt_llm/_torch/modules/kimi_kda/kimi_k3_mamba_metadata.py]: Good catch. Addressed in 640ffce03e. KDA now allocates one `(max_num_tokens, 2)` buffer on its first preparation and slices that stable storage for each batch. Because every chunk contains at least one token, `max_num_tokens` is a hard upper bound, so there is no post-warmup growth or reallocation to invalidate CUDA graph capture.

File: tests/unittest/_torch/modules/kimi_kda/test_kimi_k3_mamba_metadata.py (hunk 8 above)
Comment (jiaganc) [reply tests/unittest/_torch/modules/kimi_kda/test_kimi_k3_mamba_metadata.py]: Addressed in 640ffce03e. I removed the refresh/clear and invariant-only standalone tests, and folded the mixed context/generation coverage into the existing vendored parity table so the remaining test checks external parity rather than restating the implementation.

File: tensorrt_llm/_torch/attention_backend/interface.py [hunk 9]
```diff
@@ -347,9 +349,14 @@ def _prepare_mamba_metadata(self):
 
         if self.mamba_metadata is None:
             if isinstance(self.kv_cache_manager, BaseMambaCacheManager):
-                from ..modules.mamba.mamba2_metadata import Mamba2Metadata
-                self.mamba_metadata = Mamba2Metadata(self.max_num_requests,
-                                                     self.mamba_chunk_size)
+                metadata_cls = self.mamba_metadata_cls
+                if metadata_cls is None:
```
Comment (pengbowang-nv) [inline tensorrt_llm/_torch/attention_backend/interface.py:353]: By current code this line seems dead code?

**Review by pengbowang-nv — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/attention_backend/interface.py (hunk 9 above)
Comment (jiaganc) [reply tensorrt_llm/_torch/attention_backend/interface.py:353]: Some tests failed because of this. Seem like some tests will skip the model engine init.
